User provided rules should be marked as "Cleanup Rules" when necessary

ketkarameya opened the pull request:
User has to mark rules as "Cleanup Rules" which prevents it from being considered a "Seed Rule". A Seed rule has to be instantiable using the substitutions provided in piranha arguments. 

Moving forward, we will not allow developers to declare seed rules that are not initializable with piranha arguments. Piranha will panic in these cases. 

Note we were already following this practice in demos and also internally. 

---------------
"A Seed rule has to be instantiable using the substitutions provided in piranha arguments." means is that the provided rules in the rules.toml files, unless marked as cleanup, must either a) have no holes, or b) have holes whose names match piranha arguments. I was actually missing that case (a) was a possibility from that description, until I saw the "before" code for the test change.
While try_instantiate is being removed from rule.rs, we also had instantiate, which threw an error even before this change. That try_instantiate function was only being called from a few places, and in most of them (tests), whether it succeeded or failed was being immediately checked. However, in add_to_global_rules inside rules_store.rs, before this change, we were silently ignoring any rule for which try_instantiate failed, whereas now we panic in that case.

lazaroclapp reviewed (approved):
So, code LGTM. But, just to double-check:

1. This doesn't actually include any changes enforcing that rules are tagged "Cleanup Rule", right? It just changes the tests/examples to follow that convention?
2. The refactoring in `src/models/rule.rs` is an unrelated simplification of the code with no API/behavior change? (Should that have been a separate/unrelated PR, or am I missing something?)

ketkarameya commented:
Actually this change is "semantics altering". We now enforce that, all rules that are not marked as "cleanup rules" (in the input `toml` file) will be considered as "seed rules" and therefore should be instantiable using the input substitutions provided by the user . Basically, we will not allow un-instantiable seed rules. 
- Note I eliminated `try_instantiate`. Now we have `instantiate` only. 
- `rules.rs` is where the meat of the change is .

ketkarameya commented:
Updated the current `find_propagate` test for Java to have "holes". 
Adapted the current test case.

Added a new test case that panics because substitutions for seed rule is not provided by user.

lazaroclapp reviewed (commented):
Ok, just to make sure I understand now:

- What "A Seed rule has to be instantiable using the substitutions provided in piranha arguments." means is that the provided rules in the `rules.toml` files, unless marked as cleanup, must either a) have no holes, or b) have holes whose names match piranha arguments. I was actually missing that case (a) was a possibility from that description, until I saw the "before" code for the test change.
- While `try_instantiate` is being removed from `rule.rs`, we also had `instantiate`, which threw an error even before this change. That `try_instantiate` function was only being called from a few places, and in most of them (tests), whether it succeeded or failed was being immediately checked. However, in `add_to_global_rules` inside `rules_store.rs`, before this change, we were silently ignoring any rule for which `try_instantiate` failed, whereas now we panic in that case. Does this describe the core behavior change?

As a side note, separate from this PR, I am wondering if having to add the semantically meaningful long string "Cleanup Rule" to `groups` is a good idea here. Why not "cleanup"? Or even better, are there types available in the toml file other than string? What about an enum? What else is `groups` used for? My worry here is that the long string will be prone to typos and the fact that it's a generic string will make it harder for us to detect the source of the error (you will just get a generic "couldn't instantiate rule" panic, rather than a warning about the group name, no?).

ketkarameya commented on src/tests/test_piranha_java.rs:
```diff
@@ -65,3 +65,23 @@ create_match_tests! {
+  let piranha_arguments = piranha_arguments! {
+    path_to_codebase = path_to_codebase,
+    path_to_configurations = path_to_configurations,
+    language= JAVA.to_string(),
```
> Ok, just to make sure I understand now:
> 
> * What "A Seed rule has to be instantiable using the substitutions provided in piranha arguments." means is that the provided rules in the `rules.toml` files, unless marked as cleanup, must either a) have no holes, or b) have holes whose names match piranha arguments. I was actually missing that case (a) was a possibility from that description, until I saw the "before" code for the test change.
> * While `try_instantiate` is being removed from `rule.rs`, we also had `instantiate`, which threw an error even before this change. That `try_instantiate` function was only being called from a few places, and in most of them (tests), whether it succeeded or failed was being immediately checked. However, in `add_to_global_rules` inside `rules_store.rs`, before this change, we were silently ignoring any rule for which `try_instantiate` failed, whereas now we panic in that case. Does this describe the core behavior change?

Yes you are absolutely correct at describing the change. Yes, we were invoking `try_instantiate` and silently ignoring the unuinitializable rules. Now we dont do that. 

Yes, I agree. "Cleanup Rule" is very error-prone. I have made mistakes myself. How about having an optional boolean field `is_cleanup_rule`  (that defaults to bool).  - I'll do this in a separate PR though. 
I agree with just calling it "cleeanup" too. 

Groups cannot be enums it has to be strings (Adhoc). 
Groups are used to address a set of rules. 
The user can also create new groups when they define rules - for instance, when they want to an edge between a rule and a set of rules.

lazaroclapp replied:
> Groups are used to address a set of rules.
> The user can also create new groups when they define rules - for instance, when they want to an edge between a rule and a set of rules.

Ok, right, that makes sense. But, if being a cleanup rule has semantics beyond just being a named group, then it makes sense for it to be configured as a boolean (`is_cleanup_rule` sounds good. I assume you meant defaults to `false` above). You can additionally have a group tag automatically created for such rules, if need be, but it shouldn't be something where the user has to manually type a long string without extra checking.

lazaroclapp reviewed (approved):
Confirming this LGTM even after understanding the change 🙂
